modules/andes-core/client/src/main/java/org/wso2/andes/client/XAResource_0_9_1.java

```diff
@@ -241,7 +241,19 @@ public int prepare(Xid xid) throws XAException {
 
     @Override
     public Xid[] recover(int i) throws XAException {
-        throw new RuntimeException("Feature NotImplemented");
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("recover dtx branches with prepared state");
+        }
+
+        try {
+            List<Xid> xidList = session.recoverDtxTransactions();
+            return xidList.toArray(new Xid[xidList.size()]);
+
+        } catch (FailoverException | AMQException e) {
+            XAException xaException = new XAException("Error while recovering dtx sessions.");
+            xaException.initCause(e);
```

Comment:
XAException does not have a constructor to pass both error message and the exception. Therefore we had to do this.